pysindy/optimizers/constrained_sr3.py

```diff
@@ -0,0 +1,374 @@
+        + (0.5 / nu)\\|w-v\\|^2_2
+
+    where :math:`R(v)` is a regularization function. See the following reference
+    for more details:
+
+        Zheng, Peng, et al. "A unified framework for sparse relaxed
+        regularized regression: Sr3." IEEE Access 7 (2018): 1404-1423.
+
+    Parameters
+    ----------
+    threshold : float, optional (default 0.1)
+        Determines the strength of the regularization. When the
+        regularization function R is the l0 norm, the regularization
+        is equivalent to performing hard thresholding, and lambda
+        is chosen to threshold at the value given by this parameter.
+        This is equivalent to choosing lambda = threshold^2 / (2 * nu).
+
+    nu : float, optional (default 1)
+        Determines the level of relaxation. Decreasing nu encourages
+        w and v to be close, whereas increasing nu allows the
+        regularized coefficients v to be farther from w.
+
+    tol : float, optional (default 1e-5)
+        Tolerance used for determining convergence of the optimization
+        algorithm.
+
+    thresholder : string, optional (default 'l0')
+        Regularization function to use. Currently implemented options
+        are 'l0' (l0 norm), 'l1' (l1 norm), and 'cad' (clipped
+        absolute deviation).
+
+    max_iter : int, optional (default 30)
+        Maximum iterations of the optimization algorithm.
+
+    fit_intercept : boolean, optional (default False)
+        Whether to calculate the intercept for this model. If set to false, no
+        intercept will be used in calculations.
+
+    normalize : boolean, optional (default False)
+        This parameter is ignored when fit_intercept is set to False. If True,
+        the regressors X will be normalized before regression by subtracting
+        the mean and dividing by the l2-norm.
+
+    copy_X : boolean, optional (default True)
+        If True, X will be copied; else, it may be overwritten.
+
+    initial_guess : 2D numpy array of floats (default optional)
+        If user does not pass this, the initial guess for the optimization is
+        a naive lstsq (see below). If passes, the optimization starts
+        with this matrix as the initial starting point.
+
+    unbias : boolean, optional (default True)
+        Whether to perform an extra step of unregularized linear regression to unbias
+        the coefficients for the identified support.
+        For example, if `STLSQ(alpha=0.1)` is used then the learned coefficients will
+        be biased toward 0 due to the L2 regularization.
+        Setting `unbias=True` will trigger an additional step wherein the nonzero
+        coefficients learned by the `STLSQ` object will be updated using an
+        unregularized least-squares fit.
+
+    thresholds : 2D array of floats, optional (default None)
+        Optional threshold matrix to provide different thresholds
+        for different terms in the optimization process
+
```

Comment:
This section also needs information on the new trimming and constraint parameters.

Reply:
I added the constraint parameters but unfortunately, a lot of this code is borrowed from Kathleen and I have no idea what the trimming stuff does. Moreover, I have not tried the constrained_stlsq class (I am not sure it's functional) so I think we should just delete this from the push for now, since constrained SINDy is already working for anyone who wants to use the constrained SR3 algorithm. 

I'm not sure how you want to handle all the code from Kathleen. One option would be to chop off all the parts of Kathleen's code we don't use/understand currently and have her incorporate the rest whenever she is ready.